src/cuda/cuda_sampling.cu

```diff
@@ -431,53 +439,185 @@ __global__ void FilterOnTopP(float* scores, float* prefix_sums, float* scores_te
-  thread_top_k.Init();
   int batch = blockIdx.x;
+  int tid = threadIdx.x;
+  TopK_2 partial;
 
-  for (int i = threadIdx.x; i < vocab_size; i += blockDim.x) {
-    thread_top_k.Insert(scores_in[i + batch * vocab_size], i);
+  float const MAX_T_VAL = FLT_MAX;
+
+  for (int ite = 0; ite < k; ite++) {
+    partial.init();
+    for (auto elemId = tid; elemId < vocab_size; elemId += kBlockSize) {
+      float elem = scores_in[elemId + batch * vocab_size];
+      partial.insert(elem, elemId);
+    }
+    // reduce in thread block
+    typedef cub::BlockReduce<TopK_2, kBlockSize> BlockReduce;
+    __shared__ typename BlockReduce::TempStorage temp_storage;
+    TopK_2 top_k_sequence = BlockReduce(temp_storage).Reduce(partial, reduce_topk_op_2);
+
+    if (tid == 0) {
+      scores_out[ite + batch * k] = top_k_sequence.u / temperature;
+      indices_out[ite + batch * k] = top_k_sequence.p;
+      // set the ax value to -MAX_T_VAL so that the value doesn't get picked again
+      scores_in[batch * vocab_size + top_k_sequence.p] = -MAX_T_VAL;
+    }
+
+    __syncthreads();
   }
+}
 
-  // reduce in thread block
-  typedef cub::BlockReduce<TopK<float, max_k>, kBlockSize> BlockReduce;
+// Get top k indices and scores from unsorted input
+template <int kBlockSize>
+__global__ void GetTopKKernelDistributed(int* indices_out, float* scores_in, float* scores_out, int batch_size, 
+                                        int vocab_size, int k, float temperature, 
+                                        uint* top_k_distributed_lock, int* distributed_indices_out, float* distributed_scores_out) {
+  int batch = blockIdx.x; // batch should always be 1 for this implementation TODO: Support batch > 1 later
+  TopK_2 partial;
+
+  float const MAX_T_VAL = FLT_MAX;
+  uint tid = threadIdx.x;
+
+  int start_i = blockIdx.z * blockDim.x + tid;  
+
+  int* distributed_indices_out_curr = distributed_indices_out + blockIdx.z * k;
+  float* distributed_scores_out_curr = distributed_scores_out + blockIdx.z * k;
+
+  // Process each Threadblock's top_k
+  typedef cub::BlockReduce<TopK_2, kBlockSize> BlockReduce;
   __shared__ typename BlockReduce::TempStorage temp_storage;
-  TopK<float, max_k> top_k_sequence = BlockReduce(temp_storage).Reduce(thread_top_k, reduce_topk_op<float, max_k>);
+
+  for (int ite = 0; ite < k; ite++) {
+    partial.init();
+
+    for (auto elemId = start_i; elemId < vocab_size; elemId += kBlockSize* gridDim.z) {
+      float elem = scores_in[elemId + batch * vocab_size];
+      partial.insert(elem, elemId);
+    }
+    // reduce in thread block
+    TopK_2 top_k_sequence = BlockReduce(temp_storage).Reduce(partial, reduce_topk_op_2);
+
+    if (tid == 0) {
+      distributed_scores_out_curr[ite + batch * k] = top_k_sequence.u;
+      distributed_indices_out_curr[ite + batch * k] = top_k_sequence.p;
+
+      // set the ax value to -MAX_T_VAL so that the value doesn't get picked again
+      scores_in[batch * vocab_size + top_k_sequence.p] = -MAX_T_VAL;
```

Comment:
Needs this commit https://github.com/microsoft/onnxruntime-genai/pull/1708/commits/30dc69aaf27d0670b0a870e8a3faf0426b711954 reflected here and in the reduction logic and need to re-benchmark